In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import matplotlib as mat 
import seaborn as sns
from tabulate import tabulate

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
df = pd.read_excel("appearances.xlsx")
df

appearance_id  game_id  player_id       date          player_name  \
0     2224728_119169  2224728     119169 2012-07-13      Aron Johannsson   
1     2224732_161244  2224732     161244 2012-07-14        Conor O'Brien   
2      2224729_39467  2224729      39467 2012-07-15     Clarence Goodson   
3     2232104_119169  2232104     119169 2012-07-19      Aron Johannsson   
4      2219794_39475  2219794      39475 2012-07-22       Sacha Kljestan   
...              ...      ...        ...        ...                  ...   
3563  3415291_537467  3415291     537467 2020-09-26        Joseph Efford   
3564  3415296_367423  3415296     367423 2020-09-26         Chris Durkin   
3565  3431983_478940  3431983     478940 2020-09-26        Reggie Cannon   
3566  3450575_361104  3450575     361104 2020-09-26         Sergino Dest   
3567  3412904_124732  3412904     124732 2020-09-27  John Anthony Brooks   

     competition_id  yellow_cards  red_cards  goals  assists  minutes_played  
0               DK1             0          0      0        0              90  
1               DK1             0          0      1        0              90  
2               DK1             0          0      0        0              90  
3               ELQ             0          0      0        0              74  
4              BESC             0          0      0        0               3  
...             ...           ...        ...    ...      ...             ...  
3563            BE1             0          0      1        0              77  
3564            BE1             1          0      0        0              84  
3565            PO1             0          0      0        0              90  
3566            NL1             0          0      0        0               6  
3567             L1             1          0      0        0              90  

[3568 rows x 11 columns]

In [4]:
df.isnull().sum()

appearance_id     0
game_id           0
player_id         0
date              0
player_name       0
competition_id    0
yellow_cards      0
red_cards         0
goals             0
assists           0
minutes_played    0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3568 entries, 0 to 3567
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   appearance_id   3568 non-null   object        
 1   game_id         3568 non-null   int64         
 2   player_id       3568 non-null   int64         
 3   date            3568 non-null   datetime64[ns]
 4   player_name     3568 non-null   object        
 5   competition_id  3568 non-null   object        
 6   yellow_cards    3568 non-null   int64         
 7   red_cards       3568 non-null   int64         
 8   goals           3568 non-null   int64         
 9   assists         3568 non-null   int64         
 10  minutes_played  3568 non-null   int64         
dtypes: datetime64[ns](1), int64(7), object(3)
memory usage: 306.8+ KB


In [6]:
df.shape

(3568, 11)

In [7]:
df.columns

Index(['appearance_id', 'game_id', 'player_id', 'date', 'player_name',
       'competition_id', 'yellow_cards', 'red_cards', 'goals', 'assists',
       'minutes_played'],
      dtype='object')

In [8]:
df1 = pd.read_excel("game_events.xlsx")
df1

game_event_id       date  game_id  minute  \
0     c6a3c088ed8a38d4ce074dd73b20d3da 2012-08-19  2221641      62   
1     02d605a5c2dc4f9a6721daa583fa5405 2012-08-26  2222536      54   
2     b56c2e2e087cddb3cfe9e3d340975df9 2012-11-18  2222707      79   
3     4a15d1fff4f476f48bb60092c61641d5 2012-11-23  2222721      72   
4     daa97877f7edf2fda885b411d7197921 2013-05-17  2222782      63   
...                                ...        ...      ...     ...   
1844  4acebccbc824e45d51045d8c5e164341 2023-10-31  4194147       3   
1845  159ad5633cf9d7c3a97b593efb6c3269 2023-10-31  4194147       9   
1846  daa31f19aab26eec375884aef73c73b3 2023-10-31  4194147      65   
1847  869a13060604e769290dafe0b1f14483 2023-11-01  4194152      70   
1848  6b7ed06f13c1f67f0d7359c310f20ba0 2023-10-31  4194154      76   

               type  player_id  \
0     Substitutions       1335   
1             Cards       1321   
2     Substitutions     104203   
3     Substitutions     104203   
4             Goals     104203   
...             ...        ...   
1844          Goals     355369   
1845          Goals     355369   
1846  Substitutions     355369   
1847  Substitutions     504215   
1848  Substitutions     103064   

                                            description  player_in_id  \
0                                        , Not reported       45668.0   
1                                1. Yellow card  , Foul           NaN   
2                                            , Tactical       25740.0   
3                                            , Tactical       15583.0   
4     , Right-footed shot, 1. Goal of the Season Ass...           NaN   
...                                                 ...           ...   
1844  , Left-footed shot, 1. Tournament Goal Assist:...           NaN   
1845  , Right-footed shot, 2. Tournament Goal Assist...           NaN   
1846                                         , Tactical      544148.0   
1847                                         , Tactical      406640.0   
1848                                         , Tactical      378874.0   

      player_assist_id  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4              15077.0  
...                ...  
1844          284010.0  
1845          629588.0  
1846               NaN  
1847               NaN  
1848               NaN  

[1849 rows x 9 columns]

In [9]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1849 entries, 0 to 1848
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   game_event_id     1849 non-null   object        
 1   date              1849 non-null   datetime64[ns]
 2   game_id           1849 non-null   int64         
 3   minute            1849 non-null   int64         
 4   type              1849 non-null   object        
 5   player_id         1849 non-null   int64         
 6   description       916 non-null    object        
 7   player_in_id      1155 non-null   float64       
 8   player_assist_id  140 non-null    float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(3)
memory usage: 130.1+ KB


In [10]:
df1.shape

(1849, 9)

In [11]:
df1.columns

Index(['game_event_id', 'date', 'game_id', 'minute', 'type', 'player_id',
       'description', 'player_in_id', 'player_assist_id'],
      dtype='object')

In [12]:
df1.isnull().sum()

game_event_id          0
date                   0
game_id                0
minute                 0
type                   0
player_id              0
description          933
player_in_id         694
player_assist_id    1709
dtype: int64

In [13]:
df1["description"].value_counts()

description
, Tactical                                                                              118
, Not reported                                                                           74
, Right-footed shot, 1. Goal of the Season                                               42
1. Yellow card  , Foul                                                                   39
, Right-footed shot, 1. Tournament Goal                                                  27
                                                                                       ... 
, Right-footed shot, 2. Tournament Goal Assist: , Shot on goal, 1. Tournament Assist      1
, Right-footed shot, 3. Tournament Goal Assist: , Pass, 2. Tournament Assist              1
, Header, 1. Tournament Goal Assist: , Free kick, 1. Tournament Assist                    1
, Left-footed shot, 4. Tournament Goal Assist: , Pass, 1. Tournament Assist               1
, Left-footed shot, 1. Tournament Goal Assist: , Corner, 1. Tourname

In [14]:
df1["description"].isnull().sum()

933

In [15]:
df1["description"] = df1["description"].fillna(", Tactical")

In [16]:
df1["description"].isnull().sum()

0

In [17]:
df1["player_in_id"].isnull().sum()

694

In [18]:
df1['player_in_id'] = df1['player_in_id'].fillna(239513)

In [19]:
df1["player_in_id"].isnull().sum()

0

In [20]:
df1["player_assist_id"].isnull().sum()

1709

In [21]:
df1['player_assist_id'] = df1['player_assist_id'].fillna(82500)

In [22]:
df1["player_assist_id"].isnull().sum()

0

In [23]:
df1.isnull().sum()

game_event_id       0
date                0
game_id             0
minute              0
type                0
player_id           0
description         0
player_in_id        0
player_assist_id    0
dtype: int64

In [24]:
df1

game_event_id       date  game_id  minute  \
0     c6a3c088ed8a38d4ce074dd73b20d3da 2012-08-19  2221641      62   
1     02d605a5c2dc4f9a6721daa583fa5405 2012-08-26  2222536      54   
2     b56c2e2e087cddb3cfe9e3d340975df9 2012-11-18  2222707      79   
3     4a15d1fff4f476f48bb60092c61641d5 2012-11-23  2222721      72   
4     daa97877f7edf2fda885b411d7197921 2013-05-17  2222782      63   
...                                ...        ...      ...     ...   
1844  4acebccbc824e45d51045d8c5e164341 2023-10-31  4194147       3   
1845  159ad5633cf9d7c3a97b593efb6c3269 2023-10-31  4194147       9   
1846  daa31f19aab26eec375884aef73c73b3 2023-10-31  4194147      65   
1847  869a13060604e769290dafe0b1f14483 2023-11-01  4194152      70   
1848  6b7ed06f13c1f67f0d7359c310f20ba0 2023-10-31  4194154      76   

               type  player_id  \
0     Substitutions       1335   
1             Cards       1321   
2     Substitutions     104203   
3     Substitutions     104203   
4             Goals     104203   
...             ...        ...   
1844          Goals     355369   
1845          Goals     355369   
1846  Substitutions     355369   
1847  Substitutions     504215   
1848  Substitutions     103064   

                                            description  player_in_id  \
0                                        , Not reported       45668.0   
1                                1. Yellow card  , Foul      239513.0   
2                                            , Tactical       25740.0   
3                                            , Tactical       15583.0   
4     , Right-footed shot, 1. Goal of the Season Ass...      239513.0   
...                                                 ...           ...   
1844  , Left-footed shot, 1. Tournament Goal Assist:...      239513.0   
1845  , Right-footed shot, 2. Tournament Goal Assist...      239513.0   
1846                                         , Tactical      544148.0   
1847                                         , Tactical      406640.0   
1848                                         , Tactical      378874.0   

      player_assist_id  
0              82500.0  
1              82500.0  
2              82500.0  
3              82500.0  
4              15077.0  
...                ...  
1844          284010.0  
1845          629588.0  
1846           82500.0  
1847           82500.0  
1848           82500.0  

[1849 rows x 9 columns]

In [25]:
df2 = pd.read_excel('game_lineups.xlsx')
df2

game_lineups_id  game_id             type  number  \
0    f2570d1504fc02f4b6c7608e8dcf89a3  4087925      substitutes      34   
1    f5f0da93ea8e1d8bdd799658e7c8f7cb  4087928  starting_lineup      13   
2    31a4d12ec23d604779d909d26c1b5410  4087929      substitutes      26   
3    776dcbef98651450db76723cb7e3b4df  4087935      substitutes      26   
4    6a35ef7495303f29e7f85dbd54547fb1  4087936  starting_lineup      13   
..                                ...      ...              ...     ...   
214  ec3d266094f99ca0a8847de827e37105  4194152  starting_lineup       7   
215  667840cda9bdf3b0344b8e99b306cf38  4194152  starting_lineup      23   
216  1c5d2f60ee777760f8a757aa10c42bb1  4194154  starting_lineup      13   
217  99032084fd00ffbfee52c541a9f960ab  4204000      substitutes      14   
218  5c9eaf6ebb621d43a0d6fd6a9e607ef9  4220942      substitutes      13   

     player_id          player_name  team_captain          position  
0       242284        Ethan Horvath             0        Goalkeeper  
1       145466             Tim Ream             0       Centre-Back  
2       578539       Chris Richards             0       Centre-Back  
3       578539       Chris Richards             0       Centre-Back  
4       145466             Tim Ream             1       Centre-Back  
..         ...                  ...           ...               ...  
214     504215       Giovanni Reyna             0      Right Winger  
215     124732  John Anthony Brooks             0       Centre-Back  
216     103064        Terrence Boyd             0    Centre-Forward  
217     315762     Luca de la Torre             0  Central Midfield  
218     145466             Tim Ream             0       Centre-Back  

[219 rows x 8 columns]

In [26]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219 entries, 0 to 218
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   game_lineups_id  219 non-null    object
 1   game_id          219 non-null    int64 
 2   type             219 non-null    object
 3   number           219 non-null    int64 
 4   player_id        219 non-null    int64 
 5   player_name      219 non-null    object
 6   team_captain     219 non-null    int64 
 7   position         219 non-null    object
dtypes: int64(4), object(4)
memory usage: 13.8+ KB


In [27]:
df2.shape

(219, 8)

In [28]:
df2.isnull().sum()

game_lineups_id    0
game_id            0
type               0
number             0
player_id          0
player_name        0
team_captain       0
position           0
dtype: int64

In [29]:
df3 = pd.read_excel('games.xlsx')
df3

game_id competition_id  season         round       date  \
0     2222734            RU1    2012  19. Matchday 2012-12-07   
1     2224572            DK1    2012   3. Matchday 2012-07-28   
2     2224628            DK1    2012  22. Matchday 2013-03-08   
3     2224655            DK1    2012  22. Matchday 2013-03-10   
4     2224729            DK1    2012   1. Matchday 2012-07-15   
...       ...            ...     ...           ...        ...   
3263  3296153            NLP    2019   Round of 16 2020-01-22   
3264  2875216            DK1    2017  19. Matchday 2017-12-10   
3265  3099247            BE1    2018  11. Matchday 2018-10-20   
3266  2872273            GB1    2017  16. Matchday 2017-12-09   
3267  2517322            FAC    2014   Third Round 2015-01-04   

      home_club_goals  away_club_goals  home_club_position  \
0                   2                3                11.0   
1                   1                2                12.0   
2                   2                0                 3.0   
3                   0                3                 6.0   
4                   0                1                 9.0   
...               ...              ...                 ...   
3263                7                0                 NaN   
3264                3                2                 3.0   
3265                3                1                 9.0   
3266                5                1                 6.0   
3267                3                1                 NaN   

      away_club_position home_club_manager_name away_club_manager_name  \
0                   10.0        Miodrag Bozovic       Slavoljub Muslin   
1                    1.0            Jess Thorup       Lars Sondergaard   
2                   10.0             Colin Todd       Lars Sondergaard   
3                   12.0           Peter Rensen        Auri Skarbalius   
4                    5.0        Auri Skarbalius            Troels Bech   
...                  ...                    ...                    ...   
3263                 NaN           Erik ten Hag           Eric Meijers   
3264                 4.0        Kasper Hjulmand      Thomas Thomasberg   
3265                 7.0          Glen De Boeck              Marc Brys   
3266                15.0    Mauricio Pochettino            Mark Hughes   
3267                 NaN            Mark Hughes           Kevin Wilkin   

                   stadium  attendance              referee  \
0                  Olimp-2      5700.0     Sergey Kuznetsov   
1         Blue Water Arena      5345.0     Michael Svendsen   
2     Cepheus Park Randers      3481.0         Jakob Kehlet   
3               Ceres Park      7967.0          Kenn Hansen   
4          Brondby Stadium     10806.0  Lars Christoffersen   
...                    ...         ...                  ...   
3263   Johan Cruijff ArenA     52017.0      Richard Martens   
3264   Right to Dream Park      2843.0     Michael Tykgaard   
3265   Guldensporenstadion      6143.0          Wesley Alen   
3266       Wembley Stadium     62202.0           Roger East   
3267        bet365 Stadium     19423.0         Simon Hooper   

      home_club_formation  away_club_formation  \
0                     NaN                  NaN   
1                     NaN                  NaN   
2                     NaN                  NaN   
3                     NaN                  NaN   
4                     NaN                  NaN   
...                   ...                  ...   
3263                  NaN                  NaN   
3264                  NaN                  NaN   
3265                  NaN                  NaN   
3266                  NaN                  NaN   
3267                  NaN                  NaN   

                        home_club_name  \
0                            FK Rostov   
1                           Esbjerg fB   
2                 Randers Fodbold Club   
3            Aarhus Gymnastik Forening   
4                      

In [30]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3268 entries, 0 to 3267
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   game_id                 3268 non-null   int64         
 1   competition_id          3268 non-null   object        
 2   season                  3268 non-null   int64         
 3   round                   3268 non-null   object        
 4   date                    3268 non-null   datetime64[ns]
 5   home_club_goals         3268 non-null   int64         
 6   away_club_goals         3268 non-null   int64         
 7   home_club_position      2789 non-null   float64       
 8   away_club_position      2789 non-null   float64       
 9   home_club_manager_name  3263 non-null   object        
 10  away_club_manager_name  3263 non-null   object        
 11  stadium                 3268 non-null   object        
 12  attendance              3153 non-null   float64 

In [31]:
df3.shape

(3268, 20)

In [32]:
df3.isnull().sum()

game_id                      0
competition_id               0
season                       0
round                        0
date                         0
home_club_goals              0
away_club_goals              0
home_club_position         479
away_club_position         479
home_club_manager_name       5
away_club_manager_name       5
stadium                      0
attendance                 115
referee                      2
home_club_formation       3268
away_club_formation       3268
home_club_name             110
away_club_name              63
aggregate                    0
competition_type             0
dtype: int64

In [33]:
df3["home_club_formation"].isnull().sum()

3268

In [34]:
df3["home_club_formation"] = df3["home_club_formation"].fillna(3)

In [35]:
df3["home_club_formation"].isnull().sum()

0

In [36]:
df3["away_club_formation"].isnull().sum()

3268

In [37]:
df3["away_club_formation"] = df3["away_club_formation"].fillna(10)

In [38]:
df3["away_club_formation"].isnull().sum()

0

In [39]:
df3["home_club_manager_name"].isnull().sum()

5

In [40]:
df3["home_club_manager_name"].value_counts()

home_club_manager_name
Mark Hughes            68
Roberto Martínez       55
Hein Vanhaezebrouck    54
John van den Brom      53
Paul Lambert           53
                       ..
Morten Karlsen          1
Darko Milanic           1
Joakim Mattsson         1
John Brown              1
Luis Freire             1
Name: count, Length: 580, dtype: int64

In [41]:
df3["home_club_manager_name"] = df3["home_club_manager_name"].fillna("Mark Hughes")

In [42]:
df3["home_club_manager_name"].isnull().sum()

0

In [43]:
df3["away_club_manager_name"].isnull().sum()

5

In [44]:
df3["away_club_manager_name"].value_counts()

away_club_manager_name
Mark Hughes            71
Roberto Martínez       57
Lucien Favre           57
Paul Lambert           55
Hein Vanhaezebrouck    54
                       ..
Veaceslav Rusnac        1
Ian Murray              1
Steve Lovell            1
Raul Procopio           1
Kevin Wilkin            1
Name: count, Length: 550, dtype: int64

In [45]:
df3["away_club_manager_name"] = df3["away_club_manager_name"].fillna("Mark Hughes")

In [46]:
df3["away_club_manager_name"].isnull().sum()

0

In [47]:
df3["home_club_position"].isnull().sum()

479

In [48]:
df3["home_club_position"].value_counts()

home_club_position
3.0     202
11.0    175
4.0     174
9.0     173
5.0     169
2.0     167
10.0    165
6.0     163
7.0     161
12.0    154
8.0     140
13.0    138
14.0    137
1.0     135
17.0    126
16.0    125
15.0    110
18.0     79
19.0     51
20.0     45
Name: count, dtype: int64

In [49]:
df3["home_club_position"] = df3["home_club_position"].fillna(3)

In [50]:
df3["home_club_position"].isnull().sum()

0

In [51]:
df3["away_club_position"].isnull().sum()

479

In [52]:
df3["away_club_position"].value_counts()

away_club_position
10.0    180
6.0     177
9.0     169
11.0    165
12.0    162
2.0     159
3.0     157
1.0     154
4.0     151
8.0     150
5.0     148
7.0     146
13.0    139
17.0    133
14.0    132
16.0    125
15.0    125
18.0    107
19.0     55
20.0     54
21.0      1
Name: count, dtype: int64

In [53]:
df3["away_club_position"] = df3["away_club_position"].fillna(10)

In [54]:
df3["away_club_position"].isnull().sum()

0

In [55]:
df3["home_club_name"].isnull().sum()

110

In [56]:
df3["home_club_name"].value_counts()

home_club_name
Borussia Verein für Leibesübung 1900 e.V.    93
FC Schalke 04                                90
Stoke City                                   87
Borussia Dortmund                            87
Eintracht Frankfurt Fuayball AG              83
                                             ..
Lierse SK (- 2018)                            1
Genoa Cricket and Football Club               1
fc vitoria setubal                            1
FK Dinamo Moskva                              1
Bologna Football Club 1909                    1
Name: count, Length: 253, dtype: int64

In [57]:
df3["home_club_name"] = df3["home_club_name"].fillna("Borussia Verein für Leibesübung 1900 e.V.")

In [58]:
df3["home_club_name"].isnull().sum()

0

In [59]:
df3["away_club_name"].isnull().sum()

63

In [60]:
df3["away_club_name"].value_counts()

away_club_name
Borussia Dortmund                            95
Stoke City                                   92
eintracht frankfurt Football ag              91
FC Schalke 04                                88
Borussia Verein fur Leibesubung 1900 e.V.    87
                                             ..
FC Dordrecht                                  1
Villarreal Club de Football S.A.D             1
PFK Stal Kamyanske (-2018)                    1
Elazigspor                                    1
Desportivo Aves (- 2020)                      1
Name: count, Length: 255, dtype: int64

In [61]:
df3["away_club_name"] = df3["away_club_name"].fillna("Borussia Dortmund")

In [62]:
df3["away_club_name"].isnull().sum()

0

In [63]:
df3.isnull().sum()

game_id                     0
competition_id              0
season                      0
round                       0
date                        0
home_club_goals             0
away_club_goals             0
home_club_position          0
away_club_position          0
home_club_manager_name      0
away_club_manager_name      0
stadium                     0
attendance                115
referee                     2
home_club_formation         0
away_club_formation         0
home_club_name              0
away_club_name              0
aggregate                   0
competition_type            0
dtype: int64

In [64]:
df3["attendance"].isnull().sum()

115

In [65]:
df3["attendance"].value_counts()

attendance
42100.0    24
51500.0    22
81360.0    20
24000.0    19
30150.0    17
           ..
4360.0      1
34876.0     1
24344.0     1
19387.0     1
19423.0     1
Name: count, Length: 2604, dtype: int64

In [66]:
df3["attendance"] = df3["attendance"].fillna(42100)

In [67]:
df3["attendance"].isnull().sum()

0

In [68]:
df3["referee"].isnull().sum()

2

In [69]:
df3["referee"].value_counts()

referee
Dr. Felix Brych       60
Felix Zwayer          59
Manuel Graofe         58
Anthony Taylor        57
Martin Atkinson       57
                      ..
Kristinn Jakobsson     1
Alper Ulusoy           1
Manuel Oliveira        1
Cosme Machado          1
Antti Munukka          1
Name: count, Length: 347, dtype: int64

In [70]:
df3["referee"] = df3["referee"].fillna("Dr. Felix Brych")

In [71]:
df3["referee"].isnull().sum()

0

In [72]:
df3.isnull().sum()

game_id                   0
competition_id            0
season                    0
round                     0
date                      0
home_club_goals           0
away_club_goals           0
home_club_position        0
away_club_position        0
home_club_manager_name    0
away_club_manager_name    0
stadium                   0
attendance                0
referee                   0
home_club_formation       0
away_club_formation       0
home_club_name            0
away_club_name            0
aggregate                 0
competition_type          0
dtype: int64

In [73]:
df3

game_id competition_id  season         round       date  \
0     2222734            RU1    2012  19. Matchday 2012-12-07   
1     2224572            DK1    2012   3. Matchday 2012-07-28   
2     2224628            DK1    2012  22. Matchday 2013-03-08   
3     2224655            DK1    2012  22. Matchday 2013-03-10   
4     2224729            DK1    2012   1. Matchday 2012-07-15   
...       ...            ...     ...           ...        ...   
3263  3296153            NLP    2019   Round of 16 2020-01-22   
3264  2875216            DK1    2017  19. Matchday 2017-12-10   
3265  3099247            BE1    2018  11. Matchday 2018-10-20   
3266  2872273            GB1    2017  16. Matchday 2017-12-09   
3267  2517322            FAC    2014   Third Round 2015-01-04   

      home_club_goals  away_club_goals  home_club_position  \
0                   2                3                11.0   
1                   1                2                12.0   
2                   2                0                 3.0   
3                   0                3                 6.0   
4                   0                1                 9.0   
...               ...              ...                 ...   
3263                7                0                 3.0   
3264                3                2                 3.0   
3265                3                1                 9.0   
3266                5                1                 6.0   
3267                3                1                 3.0   

      away_club_position home_club_manager_name away_club_manager_name  \
0                   10.0        Miodrag Bozovic       Slavoljub Muslin   
1                    1.0            Jess Thorup       Lars Sondergaard   
2                   10.0             Colin Todd       Lars Sondergaard   
3                   12.0           Peter Rensen        Auri Skarbalius   
4                    5.0        Auri Skarbalius            Troels Bech   
...                  ...                    ...                    ...   
3263                10.0           Erik ten Hag           Eric Meijers   
3264                 4.0        Kasper Hjulmand      Thomas Thomasberg   
3265                 7.0          Glen De Boeck              Marc Brys   
3266                15.0    Mauricio Pochettino            Mark Hughes   
3267                10.0            Mark Hughes           Kevin Wilkin   

                   stadium  attendance              referee  \
0                  Olimp-2      5700.0     Sergey Kuznetsov   
1         Blue Water Arena      5345.0     Michael Svendsen   
2     Cepheus Park Randers      3481.0         Jakob Kehlet   
3               Ceres Park      7967.0          Kenn Hansen   
4          Brondby Stadium     10806.0  Lars Christoffersen   
...                    ...         ...                  ...   
3263   Johan Cruijff ArenA     52017.0      Richard Martens   
3264   Right to Dream Park      2843.0     Michael Tykgaard   
3265   Guldensporenstadion      6143.0          Wesley Alen   
3266       Wembley Stadium     62202.0           Roger East   
3267        bet365 Stadium     19423.0         Simon Hooper   

      home_club_formation  away_club_formation  \
0                     3.0                 10.0   
1                     3.0                 10.0   
2                     3.0                 10.0   
3                     3.0                 10.0   
4                     3.0                 10.0   
...                   ...                  ...   
3263                  3.0                 10.0   
3264                  3.0                 10.0   
3265                  3.0                 10.0   
3266                  3.0                 10.0   
3267                  3.0                 10.0   

                        home_club_name  \
0                            FK Rostov   
1                           Esbjerg fB   
2                 Randers Fodbold Club   
3            Aarhus Gymnastik Forening   
4                      

In [74]:
df4 = pd.read_excel('players.xlsx')
df4

player_id                 name  last_season  current_club_id  \
0       124732  John Anthony Brooks         2023              533   
1       223047      Emerson Hyndman         2018              903   
2       307781         Lynden Gooch         2016              289   
3       370846         Timothy Weah         2023              506   
4       484756   Djordje Mihailovic         2023             1090   
..         ...                  ...          ...              ...   
147     273570        Desevio Payne         2020             1283   
148     111783     Alejandro Bedoya         2015              995   
149     160670             Joe Gyau         2014               16   
150       3476         Brad Friedel         2014              148   
151     504215       Giovanni Reyna         2023               16   

             player_code country_of_birth date_of_birth        sub_position  \
0    john-anthony-brooks          Germany    1993-01-28         Centre-Back   
1        emerson-hyndman    United States    1996-04-09    Central Midfield   
2           lynden-gooch    United States    1995-12-24        Right Winger   
3           timothy-weah    United States    2000-02-22      Right Midfield   
4     djordje-mihailovic    United States    1998-11-10  Attacking Midfield   
..                   ...              ...           ...                 ...   
147        desevio-payne    United States    1995-11-30          Right-Back   
148     alejandro-bedoya    United States    1987-04-29    Central Midfield   
149             joe-gyau    United States    1992-09-16         Left Winger   
150         brad-friedel    United States    1971-05-18          Goalkeeper   
151       giovanni-reyna          England    2002-11-13  Attacking Midfield   

       position   foot  height_in_cm  market_value_in_eur  \
0      Defender   left         194.0            2000000.0   
1      Midfield  right         170.0             800000.0   
2        Attack  right         173.0            1000000.0   
3      Midfield  right         183.0           14000000.0   
4      Midfield    NaN         178.0            5000000.0   
..          ...    ...           ...                  ...   
147    Defender  right         182.0                  NaN   
148    Midfield  right         182.0             300000.0   
149      Attack   both         175.0             400000.0   
150  Goalkeeper    NaN           NaN                  NaN   
151    Midfield  right         185.0           25000000.0   

     highest_market_value_in_eur contract_expiration_date           agent_name  
0                     15000000.0               2024-06-30                ROGON  
1                      1800000.0                      NaT           OmniSports  
2                      1000000.0               2024-06-30  Unique Sports Group  
3                     15000000.0               2028-06-30    BS Group - BS Law  
4                      6000000.0               2026-06-30  YMU Management Ltd.  
..                           ...                      ...                  ...  
147                     400000.0                      NaT                  NaN  
148                    2500000.0               2023-12-31            YMU Group  
149                     600000.0               2023-12-31  BR Group Management  
150                    4000000.0                      NaT                  NaN  
151                   42000000.0               2025-06-30            Wasserman  

[152 rows x 15 columns]

In [75]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152 entries, 0 to 151
Data columns (total 15 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   player_id                    152 non-null    int64         
 1   name                         152 non-null    object        
 2   last_season                  152 non-null    int64         
 3   current_club_id              152 non-null    int64         
 4   player_code                  152 non-null    object        
 5   country_of_birth             152 non-null    object        
 6   date_of_birth                152 non-null    datetime64[ns]
 7   sub_position                 152 non-null    object        
 8   position                     152 non-null    object        
 9   foot                         141 non-null    object        
 10  height_in_cm                 149 non-null    float64       
 11  market_value_in_eur          114 non-null    

In [76]:
df4.shape

(152, 15)

In [77]:
df4.columns

Index(['player_id', 'name', 'last_season', 'current_club_id', 'player_code',
       'country_of_birth', 'date_of_birth', 'sub_position', 'position', 'foot',
       'height_in_cm', 'market_value_in_eur', 'highest_market_value_in_eur',
       'contract_expiration_date', 'agent_name'],
      dtype='object')

In [78]:
df4.isnull().sum()

player_id                       0
name                            0
last_season                     0
current_club_id                 0
player_code                     0
country_of_birth                0
date_of_birth                   0
sub_position                    0
position                        0
foot                           11
height_in_cm                    3
market_value_in_eur            38
highest_market_value_in_eur     4
contract_expiration_date       46
agent_name                     45
dtype: int64

In [79]:
df4["foot"].isnull().sum()

11

In [80]:
df4["foot"].value_counts()

foot
right    118
left      16
both       7
Name: count, dtype: int64

In [81]:
df4["foot"] = df4["foot"].fillna("right")

In [82]:
df4["foot"].isnull().sum()

0

In [83]:
df4["height_in_cm"].isnull().sum()

3

In [84]:
df4["height_in_cm"].value_counts()

height_in_cm
183.0    15
180.0    13
185.0    13
178.0    10
193.0    10
191.0     9
175.0     8
177.0     8
176.0     6
190.0     6
184.0     5
170.0     5
188.0     5
182.0     5
186.0     4
173.0     4
172.0     3
194.0     3
187.0     3
189.0     2
179.0     2
174.0     2
181.0     1
197.0     1
200.0     1
162.0     1
195.0     1
192.0     1
204.0     1
171.0     1
Name: count, dtype: int64

In [85]:
df4["height_in_cm"] = df4["height_in_cm"].fillna(183.0)

In [86]:
df4["height_in_cm"].isnull().sum()

0

In [87]:
df4["market_value_in_eur"].isnull().sum()

38

In [88]:
df4["market_value_in_eur"].value_counts()

market_value_in_eur
1000000.0     9
2000000.0     8
200000.0      7
400000.0      6
5000000.0     5
2500000.0     5
1500000.0     5
300000.0      5
800000.0      5
250000.0      4
150000.0      4
700000.0      4
500000.0      4
25000000.0    3
20000000.0    3
4000000.0     3
350000.0      3
10000000.0    3
8000000.0     2
100000.0      2
1200000.0     2
9000000.0     2
600000.0      2
12000000.0    2
850000.0      1
75000.0       1
650000.0      1
30000000.0    1
13000000.0    1
900000.0      1
1800000.0     1
15000000.0    1
14000000.0    1
750000.0      1
4500000.0     1
3000000.0     1
3500000.0     1
450000.0      1
125000.0      1
7000000.0     1
Name: count, dtype: int64

In [89]:
df4["market_value_in_eur"] = df4["market_value_in_eur"].fillna(1000000.0)

In [90]:
df4["market_value_in_eur"].isnull().sum()

0

In [91]:
df4["highest_market_value_in_eur"].isnull().sum()

4

In [92]:
df4["highest_market_value_in_eur"].value_counts()

highest_market_value_in_eur
3000000.0     8
1000000.0     8
2000000.0     8
4000000.0     8
10000000.0    7
400000.0      6
300000.0      6
6000000.0     5
2500000.0     5
5000000.0     5
150000.0      4
3500000.0     4
15000000.0    4
500000.0      4
250000.0      4
600000.0      4
30000000.0    3
12000000.0    3
8000000.0     3
800000.0      3
4500000.0     3
1500000.0     3
9000000.0     2
50000.0       2
1200000.0     2
25000000.0    2
900000.0      2
125000.0      2
100000.0      2
7000000.0     2
950000.0      2
700000.0      2
60000000.0    1
13000000.0    1
1400000.0     1
75000.0       1
2200000.0     1
1850000.0     1
750000.0      1
550000.0      1
1600000.0     1
850000.0      1
1700000.0     1
1800000.0     1
20000000.0    1
1750000.0     1
2800000.0     1
350000.0      1
1300000.0     1
6500000.0     1
650000.0      1
42000000.0    1
Name: count, dtype: int64

In [93]:
df4["highest_market_value_in_eur"] = df4["highest_market_value_in_eur"].fillna(3000000.0)

In [94]:
df4["highest_market_value_in_eur"].isnull().sum()

0

In [95]:
df4["contract_expiration_date"].isnull().sum()

46

In [96]:
df4["contract_expiration_date"].value_counts()

contract_expiration_date
2025-06-30    20
2024-06-30    18
2023-12-31    13
2026-06-30    11
2027-06-30    10
2024-12-31     9
2028-06-30     6
2025-12-31     6
2023-11-30     5
2023-06-30     2
2026-05-31     2
2023-05-31     1
2026-12-31     1
2027-12-31     1
2024-05-31     1
Name: count, dtype: int64

In [97]:
df4["contract_expiration_date"] = df4["contract_expiration_date"].fillna("2025-06-30")

In [98]:
df4["contract_expiration_date"].isnull().sum()

0

In [99]:
df4["agent_name"].isnull().sum()

45

In [100]:
df4["agent_name"].isnull().sum()

45

In [101]:
df4["agent_name"].value_counts()

agent_name
Wasserman               19
YMU Management Ltd.      7
CAA Stellar              7
PROSPORT Management      6
Octagon                  6
Mega Sports              5
PRO FC                   4
CCC                      3
Unique Sports Group      3
YMU Group                3
AKA Global GmbH          2
CAA Base Ltd             2
SXI                      2
SBM                      1
TrueSports GmbH          1
GROW                     1
FC Enterprise            1
Shot Sports              1
INSUA GROUP              1
Field Management         1
ROGON                    1
NVA SEG                  1
fair-sport GmbH          1
PG 120 Sport Agency      1
CT10 management Srl      1
More than Sport GmbH     1
InterStarDeporte         1
TOP Agency               1
Gestifute                1
Joes Blakborn            1
ROOF                     1
FGM                      1
Promoesport              1
ARP Sportmarketing       1
Beswicks Sports Ltd      1
BS Group - BS Law        1
Prosport         

In [102]:
df4["agent_name"] = df4["agent_name"].fillna("Wasserman")

In [103]:
df4.isnull().sum()

player_id                      0
name                           0
last_season                    0
current_club_id                0
player_code                    0
country_of_birth               0
date_of_birth                  0
sub_position                   0
position                       0
foot                           0
height_in_cm                   0
market_value_in_eur            0
highest_market_value_in_eur    0
contract_expiration_date       0
agent_name                     0
dtype: int64

In [104]:
df4

player_id                 name  last_season  current_club_id  \
0       124732  John Anthony Brooks         2023              533   
1       223047      Emerson Hyndman         2018              903   
2       307781         Lynden Gooch         2016              289   
3       370846         Timothy Weah         2023              506   
4       484756   Djordje Mihailovic         2023             1090   
..         ...                  ...          ...              ...   
147     273570        Desevio Payne         2020             1283   
148     111783     Alejandro Bedoya         2015              995   
149     160670             Joe Gyau         2014               16   
150       3476         Brad Friedel         2014              148   
151     504215       Giovanni Reyna         2023               16   

             player_code country_of_birth date_of_birth        sub_position  \
0    john-anthony-brooks          Germany    1993-01-28         Centre-Back   
1        emerson-hyndman    United States    1996-04-09    Central Midfield   
2           lynden-gooch    United States    1995-12-24        Right Winger   
3           timothy-weah    United States    2000-02-22      Right Midfield   
4     djordje-mihailovic    United States    1998-11-10  Attacking Midfield   
..                   ...              ...           ...                 ...   
147        desevio-payne    United States    1995-11-30          Right-Back   
148     alejandro-bedoya    United States    1987-04-29    Central Midfield   
149             joe-gyau    United States    1992-09-16         Left Winger   
150         brad-friedel    United States    1971-05-18          Goalkeeper   
151       giovanni-reyna          England    2002-11-13  Attacking Midfield   

       position   foot  height_in_cm  market_value_in_eur  \
0      Defender   left         194.0            2000000.0   
1      Midfield  right         170.0             800000.0   
2        Attack  right         173.0            1000000.0   
3      Midfield  right         183.0           14000000.0   
4      Midfield  right         178.0            5000000.0   
..          ...    ...           ...                  ...   
147    Defender  right         182.0            1000000.0   
148    Midfield  right         182.0             300000.0   
149      Attack   both         175.0             400000.0   
150  Goalkeeper  right         183.0            1000000.0   
151    Midfield  right         185.0           25000000.0   

     highest_market_value_in_eur contract_expiration_date           agent_name  
0                     15000000.0               2024-06-30                ROGON  
1                      1800000.0               2025-06-30           OmniSports  
2                      1000000.0               2024-06-30  Unique Sports Group  
3                     15000000.0               2028-06-30    BS Group - BS Law  
4                      6000000.0               2026-06-30  YMU Management Ltd.  
..                           ...                      ...                  ...  
147                     400000.0               2025-06-30            Wasserman  
148                    2500000.0               2023-12-31            YMU Group  
149                     600000.0               2023-12-31  BR Group Management  
150                    4000000.0               2025-06-30            Wasserman  
151                   42000000.0               2025-06-30            Wasserman  

[152 rows x 15 columns]

In [105]:
# checking the null values once

In [106]:
df.isnull().sum()

appearance_id     0
game_id           0
player_id         0
date              0
player_name       0
competition_id    0
yellow_cards      0
red_cards         0
goals             0
assists           0
minutes_played    0
dtype: int64

In [107]:
df1.isnull().sum()

game_event_id       0
date                0
game_id             0
minute              0
type                0
player_id           0
description         0
player_in_id        0
player_assist_id    0
dtype: int64

In [108]:
df2.isnull().sum()

game_lineups_id    0
game_id            0
type               0
number             0
player_id          0
player_name        0
team_captain       0
position           0
dtype: int64

In [109]:
df3.isnull().sum()

game_id                   0
competition_id            0
season                    0
round                     0
date                      0
home_club_goals           0
away_club_goals           0
home_club_position        0
away_club_position        0
home_club_manager_name    0
away_club_manager_name    0
stadium                   0
attendance                0
referee                   0
home_club_formation       0
away_club_formation       0
home_club_name            0
away_club_name            0
aggregate                 0
competition_type          0
dtype: int64

In [110]:
df4.isnull().sum()

player_id                      0
name                           0
last_season                    0
current_club_id                0
player_code                    0
country_of_birth               0
date_of_birth                  0
sub_position                   0
position                       0
foot                           0
height_in_cm                   0
market_value_in_eur            0
highest_market_value_in_eur    0
contract_expiration_date       0
agent_name                     0
dtype: int64

In [111]:
df

appearance_id  game_id  player_id       date          player_name  \
0     2224728_119169  2224728     119169 2012-07-13      Aron Johannsson   
1     2224732_161244  2224732     161244 2012-07-14        Conor O'Brien   
2      2224729_39467  2224729      39467 2012-07-15     Clarence Goodson   
3     2232104_119169  2232104     119169 2012-07-19      Aron Johannsson   
4      2219794_39475  2219794      39475 2012-07-22       Sacha Kljestan   
...              ...      ...        ...        ...                  ...   
3563  3415291_537467  3415291     537467 2020-09-26        Joseph Efford   
3564  3415296_367423  3415296     367423 2020-09-26         Chris Durkin   
3565  3431983_478940  3431983     478940 2020-09-26        Reggie Cannon   
3566  3450575_361104  3450575     361104 2020-09-26         Sergino Dest   
3567  3412904_124732  3412904     124732 2020-09-27  John Anthony Brooks   

     competition_id  yellow_cards  red_cards  goals  assists  minutes_played  
0               DK1             0          0      0        0              90  
1               DK1             0          0      1        0              90  
2               DK1             0          0      0        0              90  
3               ELQ             0          0      0        0              74  
4              BESC             0          0      0        0               3  
...             ...           ...        ...    ...      ...             ...  
3563            BE1             0          0      1        0              77  
3564            BE1             1          0      0        0              84  
3565            PO1             0          0      0        0              90  
3566            NL1             0          0      0        0               6  
3567             L1             1          0      0        0              90  

[3568 rows x 11 columns]

In [112]:
df1

game_event_id       date  game_id  minute  \
0     c6a3c088ed8a38d4ce074dd73b20d3da 2012-08-19  2221641      62   
1     02d605a5c2dc4f9a6721daa583fa5405 2012-08-26  2222536      54   
2     b56c2e2e087cddb3cfe9e3d340975df9 2012-11-18  2222707      79   
3     4a15d1fff4f476f48bb60092c61641d5 2012-11-23  2222721      72   
4     daa97877f7edf2fda885b411d7197921 2013-05-17  2222782      63   
...                                ...        ...      ...     ...   
1844  4acebccbc824e45d51045d8c5e164341 2023-10-31  4194147       3   
1845  159ad5633cf9d7c3a97b593efb6c3269 2023-10-31  4194147       9   
1846  daa31f19aab26eec375884aef73c73b3 2023-10-31  4194147      65   
1847  869a13060604e769290dafe0b1f14483 2023-11-01  4194152      70   
1848  6b7ed06f13c1f67f0d7359c310f20ba0 2023-10-31  4194154      76   

               type  player_id  \
0     Substitutions       1335   
1             Cards       1321   
2     Substitutions     104203   
3     Substitutions     104203   
4             Goals     104203   
...             ...        ...   
1844          Goals     355369   
1845          Goals     355369   
1846  Substitutions     355369   
1847  Substitutions     504215   
1848  Substitutions     103064   

                                            description  player_in_id  \
0                                        , Not reported       45668.0   
1                                1. Yellow card  , Foul      239513.0   
2                                            , Tactical       25740.0   
3                                            , Tactical       15583.0   
4     , Right-footed shot, 1. Goal of the Season Ass...      239513.0   
...                                                 ...           ...   
1844  , Left-footed shot, 1. Tournament Goal Assist:...      239513.0   
1845  , Right-footed shot, 2. Tournament Goal Assist...      239513.0   
1846                                         , Tactical      544148.0   
1847                                         , Tactical      406640.0   
1848                                         , Tactical      378874.0   

      player_assist_id  
0              82500.0  
1              82500.0  
2              82500.0  
3              82500.0  
4              15077.0  
...                ...  
1844          284010.0  
1845          629588.0  
1846           82500.0  
1847           82500.0  
1848           82500.0  

[1849 rows x 9 columns]

In [113]:
df2

game_lineups_id  game_id             type  number  \
0    f2570d1504fc02f4b6c7608e8dcf89a3  4087925      substitutes      34   
1    f5f0da93ea8e1d8bdd799658e7c8f7cb  4087928  starting_lineup      13   
2    31a4d12ec23d604779d909d26c1b5410  4087929      substitutes      26   
3    776dcbef98651450db76723cb7e3b4df  4087935      substitutes      26   
4    6a35ef7495303f29e7f85dbd54547fb1  4087936  starting_lineup      13   
..                                ...      ...              ...     ...   
214  ec3d266094f99ca0a8847de827e37105  4194152  starting_lineup       7   
215  667840cda9bdf3b0344b8e99b306cf38  4194152  starting_lineup      23   
216  1c5d2f60ee777760f8a757aa10c42bb1  4194154  starting_lineup      13   
217  99032084fd00ffbfee52c541a9f960ab  4204000      substitutes      14   
218  5c9eaf6ebb621d43a0d6fd6a9e607ef9  4220942      substitutes      13   

     player_id          player_name  team_captain          position  
0       242284        Ethan Horvath             0        Goalkeeper  
1       145466             Tim Ream             0       Centre-Back  
2       578539       Chris Richards             0       Centre-Back  
3       578539       Chris Richards             0       Centre-Back  
4       145466             Tim Ream             1       Centre-Back  
..         ...                  ...           ...               ...  
214     504215       Giovanni Reyna             0      Right Winger  
215     124732  John Anthony Brooks             0       Centre-Back  
216     103064        Terrence Boyd             0    Centre-Forward  
217     315762     Luca de la Torre             0  Central Midfield  
218     145466             Tim Ream             0       Centre-Back  

[219 rows x 8 columns]

In [114]:
df3

game_id competition_id  season         round       date  \
0     2222734            RU1    2012  19. Matchday 2012-12-07   
1     2224572            DK1    2012   3. Matchday 2012-07-28   
2     2224628            DK1    2012  22. Matchday 2013-03-08   
3     2224655            DK1    2012  22. Matchday 2013-03-10   
4     2224729            DK1    2012   1. Matchday 2012-07-15   
...       ...            ...     ...           ...        ...   
3263  3296153            NLP    2019   Round of 16 2020-01-22   
3264  2875216            DK1    2017  19. Matchday 2017-12-10   
3265  3099247            BE1    2018  11. Matchday 2018-10-20   
3266  2872273            GB1    2017  16. Matchday 2017-12-09   
3267  2517322            FAC    2014   Third Round 2015-01-04   

      home_club_goals  away_club_goals  home_club_position  \
0                   2                3                11.0   
1                   1                2                12.0   
2                   2                0                 3.0   
3                   0                3                 6.0   
4                   0                1                 9.0   
...               ...              ...                 ...   
3263                7                0                 3.0   
3264                3                2                 3.0   
3265                3                1                 9.0   
3266                5                1                 6.0   
3267                3                1                 3.0   

      away_club_position home_club_manager_name away_club_manager_name  \
0                   10.0        Miodrag Bozovic       Slavoljub Muslin   
1                    1.0            Jess Thorup       Lars Sondergaard   
2                   10.0             Colin Todd       Lars Sondergaard   
3                   12.0           Peter Rensen        Auri Skarbalius   
4                    5.0        Auri Skarbalius            Troels Bech   
...                  ...                    ...                    ...   
3263                10.0           Erik ten Hag           Eric Meijers   
3264                 4.0        Kasper Hjulmand      Thomas Thomasberg   
3265                 7.0          Glen De Boeck              Marc Brys   
3266                15.0    Mauricio Pochettino            Mark Hughes   
3267                10.0            Mark Hughes           Kevin Wilkin   

                   stadium  attendance              referee  \
0                  Olimp-2      5700.0     Sergey Kuznetsov   
1         Blue Water Arena      5345.0     Michael Svendsen   
2     Cepheus Park Randers      3481.0         Jakob Kehlet   
3               Ceres Park      7967.0          Kenn Hansen   
4          Brondby Stadium     10806.0  Lars Christoffersen   
...                    ...         ...                  ...   
3263   Johan Cruijff ArenA     52017.0      Richard Martens   
3264   Right to Dream Park      2843.0     Michael Tykgaard   
3265   Guldensporenstadion      6143.0          Wesley Alen   
3266       Wembley Stadium     62202.0           Roger East   
3267        bet365 Stadium     19423.0         Simon Hooper   

      home_club_formation  away_club_formation  \
0                     3.0                 10.0   
1                     3.0                 10.0   
2                     3.0                 10.0   
3                     3.0                 10.0   
4                     3.0                 10.0   
...                   ...                  ...   
3263                  3.0                 10.0   
3264                  3.0                 10.0   
3265                  3.0                 10.0   
3266                  3.0                 10.0   
3267                  3.0                 10.0   

                        home_club_name  \
0                            FK Rostov   
1                           Esbjerg fB   
2                 Randers Fodbold Club   
3            Aarhus Gymnastik Forening   
4                      

In [115]:
df4

player_id                 name  last_season  current_club_id  \
0       124732  John Anthony Brooks         2023              533   
1       223047      Emerson Hyndman         2018              903   
2       307781         Lynden Gooch         2016              289   
3       370846         Timothy Weah         2023              506   
4       484756   Djordje Mihailovic         2023             1090   
..         ...                  ...          ...              ...   
147     273570        Desevio Payne         2020             1283   
148     111783     Alejandro Bedoya         2015              995   
149     160670             Joe Gyau         2014               16   
150       3476         Brad Friedel         2014              148   
151     504215       Giovanni Reyna         2023               16   

             player_code country_of_birth date_of_birth        sub_position  \
0    john-anthony-brooks          Germany    1993-01-28         Centre-Back   
1        emerson-hyndman    United States    1996-04-09    Central Midfield   
2           lynden-gooch    United States    1995-12-24        Right Winger   
3           timothy-weah    United States    2000-02-22      Right Midfield   
4     djordje-mihailovic    United States    1998-11-10  Attacking Midfield   
..                   ...              ...           ...                 ...   
147        desevio-payne    United States    1995-11-30          Right-Back   
148     alejandro-bedoya    United States    1987-04-29    Central Midfield   
149             joe-gyau    United States    1992-09-16         Left Winger   
150         brad-friedel    United States    1971-05-18          Goalkeeper   
151       giovanni-reyna          England    2002-11-13  Attacking Midfield   

       position   foot  height_in_cm  market_value_in_eur  \
0      Defender   left         194.0            2000000.0   
1      Midfield  right         170.0             800000.0   
2        Attack  right         173.0            1000000.0   
3      Midfield  right         183.0           14000000.0   
4      Midfield  right         178.0            5000000.0   
..          ...    ...           ...                  ...   
147    Defender  right         182.0            1000000.0   
148    Midfield  right         182.0             300000.0   
149      Attack   both         175.0             400000.0   
150  Goalkeeper  right         183.0            1000000.0   
151    Midfield  right         185.0           25000000.0   

     highest_market_value_in_eur contract_expiration_date           agent_name  
0                     15000000.0               2024-06-30                ROGON  
1                      1800000.0               2025-06-30           OmniSports  
2                      1000000.0               2024-06-30  Unique Sports Group  
3                     15000000.0               2028-06-30    BS Group - BS Law  
4                      6000000.0               2026-06-30  YMU Management Ltd.  
..                           ...                      ...                  ...  
147                     400000.0               2025-06-30            Wasserman  
148                    2500000.0               2023-12-31            YMU Group  
149                     600000.0               2023-12-31  BR Group Management  
150                    4000000.0               2025-06-30            Wasserman  
151                   42000000.0               2025-06-30            Wasserman  

[152 rows x 15 columns]

# Merging the data together

In [116]:
# appearances and players 
# df - appearances, df1 = game_events, df2 = game_lineups, df3 = games, df4 = players. 

app_merge = pd.merge(df,df4, on = "player_id", how = "left")
app_merge.reset_index(drop = True, inplace = True)
app_merge

appearance_id  game_id  player_id       date          player_name  \
0     2224728_119169  2224728     119169 2012-07-13      Aron Johannsson   
1     2224732_161244  2224732     161244 2012-07-14        Conor O'Brien   
2      2224729_39467  2224729      39467 2012-07-15     Clarence Goodson   
3     2232104_119169  2232104     119169 2012-07-19      Aron Johannsson   
4      2219794_39475  2219794      39475 2012-07-22       Sacha Kljestan   
...              ...      ...        ...        ...                  ...   
3563  3415291_537467  3415291     537467 2020-09-26        Joseph Efford   
3564  3415296_367423  3415296     367423 2020-09-26         Chris Durkin   
3565  3431983_478940  3431983     478940 2020-09-26        Reggie Cannon   
3566  3450575_361104  3450575     361104 2020-09-26         Sergino Dest   
3567  3412904_124732  3412904     124732 2020-09-27  John Anthony Brooks   

     competition_id  yellow_cards  red_cards  goals  assists  ...  \
0               DK1             0          0      0        0  ...   
1               DK1             0          0      1        0  ...   
2               DK1             0          0      0        0  ...   
3               ELQ             0          0      0        0  ...   
4              BESC             0          0      0        0  ...   
...             ...           ...        ...    ...      ...  ...   
3563            BE1             0          0      1        0  ...   
3564            BE1             1          0      0        0  ...   
3565            PO1             0          0      0        0  ...   
3566            NL1             0          0      0        0  ...   
3567             L1             1          0      0        0  ...   

      country_of_birth date_of_birth        sub_position  position   foot  \
0        United States    1990-11-10      Centre-Forward    Attack  right   
1        United States    1988-10-20  Defensive Midfield  Midfield  right   
2        United States    1982-05-17         Centre-Back  Defender  right   
3        United States    1990-11-10      Centre-Forward    Attack  right   
4        United States    1985-09-09  Attacking Midfield  Midfield   both   
...                ...           ...                 ...       ...    ...   
3563     United States    1996-08-29      Centre-Forward    Attack  right   
3564     United States    2000-02-08  Defensive Midfield  Midfield  right   
3565     United States    1998-06-11          Right-Back  Defender  right   
3566       Netherlands    2000-11-03          Right-Back  Defender  right   
3567           Germany    1993-01-28         Centre-Back  Defender   left   

     height_in_cm market_value_in_eur highest_market_value_in_eur  \
0           184.0            400000.0                   4000000.0   
1           177.0           1000000.0                    750000.0   
2           193.0           1000000.0                   1850000.0   
3           184.0            400000.0                   4000000.0   
4           185.0           1000000.0                   3000000.0   
...           ...                 ...                         ...   
3563        176.0            300000.0                    600000.0   
3564        182.0           1200000.0                   1200000.0   
3565        180.0           4000000.0                   4000000.0   
3566        171.0          10000000.0                  30000000.0   
3567        194.0           2000000.0                  15000000.0   

     contract_expiration_date           agent_name  
0                  2024-12-31          CAA Stellar  
1                  2025-06-30            Wasserman  
2                  2025-06-30            Wasserman  
3                  2024-12-31          CAA Stellar  
4                  2025-06-30            Wasserman  
...                       ...                  ...  
3563               2026-06-30           TOP Agency  
3564               2024-12-31          Mega Sports  
3565               2025-06-30  YMU

In [117]:
# merging game events using the "game_id" which is in both columns 
# # df - appearances, df1 = game_events, df2 = game_lineups, df3 = games, df4 = players. app_merge = appearances and players

events_merge = pd.merge(app_merge, df1, on = "game_id", how = "left")
events_merge.reset_index(inplace = True)
events_merge

index   appearance_id  game_id  player_id_x     date_x  \
0         0  2224728_119169  2224728       119169 2012-07-13   
1         1  2224732_161244  2224732       161244 2012-07-14   
2         2   2224729_39467  2224729        39467 2012-07-15   
3         3  2232104_119169  2232104       119169 2012-07-19   
4         4   2219794_39475  2219794        39475 2012-07-22   
...     ...             ...      ...          ...        ...   
3804   3804  3415291_537467  3415291       537467 2020-09-26   
3805   3805  3415296_367423  3415296       367423 2020-09-26   
3806   3806  3431983_478940  3431983       478940 2020-09-26   
3807   3807  3450575_361104  3450575       361104 2020-09-26   
3808   3808  3412904_124732  3412904       124732 2020-09-27   

              player_name competition_id  yellow_cards  red_cards  goals  ...  \
0         Aron Johannsson            DK1             0          0      0  ...   
1           Conor O'Brien            DK1             0          0      1  ...   
2        Clarence Goodson            DK1             0          0      0  ...   
3         Aron Johannsson            ELQ             0          0      0  ...   
4          Sacha Kljestan           BESC             0          0      0  ...   
...                   ...            ...           ...        ...    ...  ...   
3804        Joseph Efford            BE1             0          0      1  ...   
3805         Chris Durkin            BE1             1          0      0  ...   
3806        Reggie Cannon            PO1             0          0      0  ...   
3807         Sergino Dest            NL1             0          0      0  ...   
3808  John Anthony Brooks             L1             1          0      0  ...   

      contract_expiration_date           agent_name  \
0                   2024-12-31          CAA Stellar   
1                   2025-06-30            Wasserman   
2                   2025-06-30            Wasserman   
3                   2024-12-31          CAA Stellar   
4                   2025-06-30            Wasserman   
...                        ...                  ...   
3804                2026-06-30           TOP Agency   
3805                2024-12-31          Mega Sports   
3806                2025-06-30  YMU Management Ltd.   
3807                2024-06-30        Joes Blakborn   
3808                2024-06-30                ROGON   

                         game_event_id     date_y  minute           type  \
0                                  NaN        NaT     NaN            NaN   
1     d9f02a9d767f319e58909ba1e1de6d23 2012-07-14    50.0          Goals   
2                                  NaN        NaT     NaN            NaN   
3     2ab0ea7af8313c0da07c723728bb6a49 2012-07-19    74.0  Substitutions   
4                                  NaN        NaT     NaN            NaN   
...                                ...        ...     ...            ...   
3804  2c1bc4101fb8b426e84b9b5d55eb5783 2020-09-26    77.0  Substitutions   
3805  b3d9d733681d01939841c363789f9642 2020-09-26    84.0  Substitutions   
3806                               NaN        NaT     NaN            NaN   
3807                               NaN        NaT     NaN            NaN   
3808                               NaN        NaT     NaN            NaN   

     player_id_y                                        description  \
0            NaN                                                NaN   
1       161244.0  , Long distance kick, 1. Goal of the Season As...   
2            NaN                                                NaN   
3       119169.0                                     , Not reported   
4            NaN                                                NaN   
...          ...                                                ...   
3804    537467.0                                         , Tactical   
3805    367423.0                                         , Tactical   
3806         NaN                                   

In [118]:
# merging game lineups with its common column 
# # df - appearances, df1 = game_events, df2 = game_lineups, df3 = games, df4 = players.
# app_merge = appearances and players, events_merge = appearances, players and game_events 

lineups_merge = pd.merge(app_merge, df2, on = "game_id", how = "left")
lineups_merge.reset_index(drop = True, inplace = True)
lineups_merge

appearance_id  game_id  player_id_x       date        player_name_x  \
0     2224728_119169  2224728       119169 2012-07-13      Aron Johannsson   
1     2224732_161244  2224732       161244 2012-07-14        Conor O'Brien   
2      2224729_39467  2224729        39467 2012-07-15     Clarence Goodson   
3     2232104_119169  2232104       119169 2012-07-19      Aron Johannsson   
4      2219794_39475  2219794        39475 2012-07-22       Sacha Kljestan   
...              ...      ...          ...        ...                  ...   
3563  3415291_537467  3415291       537467 2020-09-26        Joseph Efford   
3564  3415296_367423  3415296       367423 2020-09-26         Chris Durkin   
3565  3431983_478940  3431983       478940 2020-09-26        Reggie Cannon   
3566  3450575_361104  3450575       361104 2020-09-26         Sergino Dest   
3567  3412904_124732  3412904       124732 2020-09-27  John Anthony Brooks   

     competition_id  yellow_cards  red_cards  goals  assists  ...  \
0               DK1             0          0      0        0  ...   
1               DK1             0          0      1        0  ...   
2               DK1             0          0      0        0  ...   
3               ELQ             0          0      0        0  ...   
4              BESC             0          0      0        0  ...   
...             ...           ...        ...    ...      ...  ...   
3563            BE1             0          0      1        0  ...   
3564            BE1             1          0      0        0  ...   
3565            PO1             0          0      0        0  ...   
3566            NL1             0          0      0        0  ...   
3567             L1             1          0      0        0  ...   

      highest_market_value_in_eur contract_expiration_date  \
0                       4000000.0               2024-12-31   
1                        750000.0               2025-06-30   
2                       1850000.0               2025-06-30   
3                       4000000.0               2024-12-31   
4                       3000000.0               2025-06-30   
...                           ...                      ...   
3563                     600000.0               2026-06-30   
3564                    1200000.0               2024-12-31   
3565                    4000000.0               2025-06-30   
3566                   30000000.0               2024-06-30   
3567                   15000000.0               2024-06-30   

               agent_name  game_lineups_id type number player_id_y  \
0             CAA Stellar              NaN  NaN    NaN         NaN   
1               Wasserman              NaN  NaN    NaN         NaN   
2               Wasserman              NaN  NaN    NaN         NaN   
3             CAA Stellar              NaN  NaN    NaN         NaN   
4               Wasserman              NaN  NaN    NaN         NaN   
...                   ...              ...  ...    ...         ...   
3563           TOP Agency              NaN  NaN    NaN         NaN   
3564          Mega Sports              NaN  NaN    NaN         NaN   
3565  YMU Management Ltd.              NaN  NaN    NaN         NaN   
3566        Joes Blakborn              NaN  NaN    NaN         NaN   
3567                ROGON              NaN  NaN    NaN         NaN   

     player_name_y team_captain position_y  
0              NaN          NaN        NaN  
1              NaN          NaN        NaN  
2              NaN          NaN        NaN  
3              NaN          NaN        NaN  
4              NaN          NaN        NaN  
...            ...          ...        ...  
3563           NaN          NaN        NaN  
3564           NaN          NaN        NaN  
3565           NaN          NaN        NaN  
3566           NaN          NaN        NaN  
3567           NaN          NaN        NaN  

[3568 rows x 32 columns]

In [119]:
# Merging the games column 
# df - appearances, df1 = game_events, df2 = game_lineups, df3 = games, df4 = players.
# app_merge = appearances and players, events_merge = appearances, players and game_events and 
# lineups_merge = appearances, players, game_event and game_lineups

Final_data = pd.merge(lineups_merge, df3, on = "game_id", how = "left")
Final_data.reset_index(drop = True, inplace = True)
Final_data

appearance_id  game_id  player_id_x     date_x        player_name_x  \
0     2224728_119169  2224728       119169 2012-07-13      Aron Johannsson   
1     2224732_161244  2224732       161244 2012-07-14        Conor O'Brien   
2      2224729_39467  2224729        39467 2012-07-15     Clarence Goodson   
3     2232104_119169  2232104       119169 2012-07-19      Aron Johannsson   
4      2219794_39475  2219794        39475 2012-07-22       Sacha Kljestan   
...              ...      ...          ...        ...                  ...   
3563  3415291_537467  3415291       537467 2020-09-26        Joseph Efford   
3564  3415296_367423  3415296       367423 2020-09-26         Chris Durkin   
3565  3431983_478940  3431983       478940 2020-09-26        Reggie Cannon   
3566  3450575_361104  3450575       361104 2020-09-26         Sergino Dest   
3567  3412904_124732  3412904       124732 2020-09-27  John Anthony Brooks   

     competition_id_x  yellow_cards  red_cards  goals  assists  ...  \
0                 DK1             0          0      0        0  ...   
1                 DK1             0          0      1        0  ...   
2                 DK1             0          0      0        0  ...   
3                 ELQ             0          0      0        0  ...   
4                BESC             0          0      0        0  ...   
...               ...           ...        ...    ...      ...  ...   
3563              BE1             0          0      1        0  ...   
3564              BE1             1          0      0        0  ...   
3565              PO1             0          0      0        0  ...   
3566              NL1             0          0      0        0  ...   
3567               L1             1          0      0        0  ...   

      away_club_manager_name                        stadium  attendance  \
0               Kent Nielsen                     Ceres Park      9728.0   
1                 Colin Todd                   Sydbank Park      2995.0   
2                Troels Bech                Brondby Stadium     10806.0   
3        Temur Shalamberidze                     Ceres Park      9000.0   
4                 Peter Maes  Constant Vanden Stock Stadion     14485.0   
...                      ...                            ...         ...   
3563             Nicky Hayen              Olympisch Stadion      3200.0   
3564            Kevin Muscat                   AFAS Stadion      5185.0   
3565        Sergio Conceicao               Estádio do Bessa     42100.0   
3566           Thomas Letsch            Johan Cruijff ArenA     15000.0   
3567          Oliver Glasner                 Dreisamstadion      3200.0   

                  referee home_club_formation away_club_formation  \
0        Michael Svendsen                 3.0                10.0   
1         Claus Bo Larsen                 3.0                10.0   
2     Lars Christoffersen                 3.0                10.0   
3         Boako Jovanetic                 3.0                10.0   
4       Laurent Colemonts                 3.0                10.0   
...                   ...                 ...                 ...   
3563      Wesli De Cremer                 3.0                10.0   
3564     Nathan Verboomen                 3.0                10.0   
3565         Luís Godinho                 3.0                10.0   
3566       Pol van Boekel                 3.0                10.0   
3567         Soren Storks                 3.0                10.0   

                                   home_club_name  \
0                       Aarhus Gymnastik Forening   
1                                  Sonderjyske Fc   
2                                      Brondby IF   
3                       Aarhus Gymnastik Forening   
4                  Royal Sporting Club Anderlecht   
...                                           ...   
3563                               Beerschot V.A.   
3564  Yellow-Red Koninklijke Voetbalclub Mechelen   
3565      

In [120]:
Final_data.isnull().sum()

appearance_id                     0
game_id                           0
player_id_x                       0
date_x                            0
player_name_x                     0
competition_id_x                  0
yellow_cards                      0
red_cards                         0
goals                             0
assists                           0
minutes_played                    0
name                              0
last_season                       0
current_club_id                   0
player_code                       0
country_of_birth                  0
date_of_birth                     0
sub_position                      0
position_x                        0
foot                              0
height_in_cm                      0
market_value_in_eur               0
highest_market_value_in_eur       0
contract_expiration_date          0
agent_name                        0
game_lineups_id                3568
type                           3568
number                      

game_lineups_id                3568
type                           3568
number                         3568
player_id_y                    3568
player_name_y                  3568
team_captain                   3568
position_y                     3568
home_club_formation               0
away_club_formation               0

In [121]:
Final_Data = Final_data.drop(columns = ["game_lineups_id","type","number","player_id_y","player_name_y","team_captain",
                                        "position_y","home_club_formation","away_club_formation","date_y","competition_id_y",
                                       "game_id","appearance_id","player_id_x"])

In [122]:
Final_Data

date_x        player_name_x competition_id_x  yellow_cards  \
0    2012-07-13      Aron Johannsson              DK1             0   
1    2012-07-14        Conor O'Brien              DK1             0   
2    2012-07-15     Clarence Goodson              DK1             0   
3    2012-07-19      Aron Johannsson              ELQ             0   
4    2012-07-22       Sacha Kljestan             BESC             0   
...         ...                  ...              ...           ...   
3563 2020-09-26        Joseph Efford              BE1             0   
3564 2020-09-26         Chris Durkin              BE1             1   
3565 2020-09-26        Reggie Cannon              PO1             0   
3566 2020-09-26         Sergino Dest              NL1             0   
3567 2020-09-27  John Anthony Brooks               L1             1   

      red_cards  goals  assists  minutes_played                 name  \
0             0      0        0              90      Aron Johannsson   
1             0      1        0              90        Conor O'Brien   
2             0      0        0              90     Clarence Goodson   
3             0      0        0              74      Aron Johannsson   
4             0      0        0               3       Sacha Kljestan   
...         ...    ...      ...             ...                  ...   
3563          0      1        0              77        Joseph Efford   
3564          0      0        0              84         Chris Durkin   
3565          0      0        0              90        Reggie Cannon   
3566          0      0        0               6         Sergino Dest   
3567          0      0        0              90  John Anthony Brooks   

      last_season  ...  away_club_position home_club_manager_name  \
0            2018  ...                 7.0           Peter Rensen   
1            2016  ...                12.0       Lars Sondergaard   
2            2012  ...                 5.0        Auri Skarbalius   
3            2018  ...                10.0           Peter Rensen   
4            2014  ...                10.0      John van den Brom   
...           ...  ...                 ...                    ...   
3563         2023  ...                18.0          Hernan Losada   
3564         2021  ...                16.0        Wouter Vrancken   
3565         2022  ...                 1.0           Vasco Seabra   
3566         2023  ...                 6.0           Erik ten Hag   
3567         2023  ...                13.0      Christian Streich   

     away_club_manager_name                        stadium attendance  \
0              Kent Nielsen                     Ceres Park     9728.0   
1                Colin Todd                   Sydbank Park     2995.0   
2               Troels Bech                Brondby Stadium    10806.0   
3       Temur Shalamberidze                     Ceres Park     9000.0   
4                Peter Maes  Constant Vanden Stock Stadion    14485.0   
...                     ...                            ...        ...   
3563            Nicky Hayen              Olympisch Stadion     3200.0   
3564           Kevin Muscat                   AFAS Stadion     5185.0   
3565       Sergio Conceicao               Estádio do Bessa    42100.0   
3566          Thomas Letsch            Johan Cruijff ArenA    15000.0   
3567         Oliver Glasner                 Dreisamstadion     3200.0   

                  referee                               home_club_name  \
0        Michael Svendsen                    Aarhus Gymnastik Forening   
1         Claus Bo Larsen                               Sonderjyske Fc   
2     Lars Christoffersen                                   Brondby IF   
3         Boako Jovanetic                    Aarhus Gymnastik Forening   
4       Laurent Colemonts               Royal Sporting Club Anderlecht   
...                   ...                                          ...   
3563      Wesli De Cremer                               Beerschot V

In [123]:
Final_Data.isnull().sum()

date_x                         0
player_name_x                  0
competition_id_x               0
yellow_cards                   0
red_cards                      0
goals                          0
assists                        0
minutes_played                 0
name                           0
last_season                    0
current_club_id                0
player_code                    0
country_of_birth               0
date_of_birth                  0
sub_position                   0
position_x                     0
foot                           0
height_in_cm                   0
market_value_in_eur            0
highest_market_value_in_eur    0
contract_expiration_date       0
agent_name                     0
season                         0
round                          0
home_club_goals                0
away_club_goals                0
home_club_position             0
away_club_position             0
home_club_manager_name         0
away_club_manager_name         0
stadium   

In [124]:
Final_Data.columns

Index(['date_x', 'player_name_x', 'competition_id_x', 'yellow_cards',
       'red_cards', 'goals', 'assists', 'minutes_played', 'name',
       'last_season', 'current_club_id', 'player_code', 'country_of_birth',
       'date_of_birth', 'sub_position', 'position_x', 'foot', 'height_in_cm',
       'market_value_in_eur', 'highest_market_value_in_eur',
       'contract_expiration_date', 'agent_name', 'season', 'round',
       'home_club_goals', 'away_club_goals', 'home_club_position',
       'away_club_position', 'home_club_manager_name',
       'away_club_manager_name', 'stadium', 'attendance', 'referee',
       'home_club_name', 'away_club_name', 'aggregate', 'competition_type'],
      dtype='object')

In [125]:
Final_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3568 entries, 0 to 3567
Data columns (total 37 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   date_x                       3568 non-null   datetime64[ns]
 1   player_name_x                3568 non-null   object        
 2   competition_id_x             3568 non-null   object        
 3   yellow_cards                 3568 non-null   int64         
 4   red_cards                    3568 non-null   int64         
 5   goals                        3568 non-null   int64         
 6   assists                      3568 non-null   int64         
 7   minutes_played               3568 non-null   int64         
 8   name                         3568 non-null   object        
 9   last_season                  3568 non-null   int64         
 10  current_club_id              3568 non-null   int64         
 11  player_code                  3568 non-null 

In [126]:
Final_Data.shape

(3568, 37)

In [127]:
Final_Data.isnull().sum()

date_x                         0
player_name_x                  0
competition_id_x               0
yellow_cards                   0
red_cards                      0
goals                          0
assists                        0
minutes_played                 0
name                           0
last_season                    0
current_club_id                0
player_code                    0
country_of_birth               0
date_of_birth                  0
sub_position                   0
position_x                     0
foot                           0
height_in_cm                   0
market_value_in_eur            0
highest_market_value_in_eur    0
contract_expiration_date       0
agent_name                     0
season                         0
round                          0
home_club_goals                0
away_club_goals                0
home_club_position             0
away_club_position             0
home_club_manager_name         0
away_club_manager_name         0
stadium   

In [128]:
#numerical columns
Num = Final_Data.select_dtypes(exclude = 'object')
Num

date_x  yellow_cards  red_cards  goals  assists  minutes_played  \
0    2012-07-13             0          0      0        0              90   
1    2012-07-14             0          0      1        0              90   
2    2012-07-15             0          0      0        0              90   
3    2012-07-19             0          0      0        0              74   
4    2012-07-22             0          0      0        0               3   
...         ...           ...        ...    ...      ...             ...   
3563 2020-09-26             0          0      1        0              77   
3564 2020-09-26             1          0      0        0              84   
3565 2020-09-26             0          0      0        0              90   
3566 2020-09-26             0          0      0        0               6   
3567 2020-09-27             1          0      0        0              90   

      last_season  current_club_id date_of_birth  height_in_cm  \
0            2018               86    1990-11-10         184.0   
1            2016             2414    1988-10-20         177.0   
2            2012              206    1982-05-17         193.0   
3            2018               86    1990-11-10         184.0   
4            2014               58    1985-09-09         185.0   
...           ...              ...           ...           ...   
3563         2023             2671    1996-08-29         176.0   
3564         2021              475    2000-02-08         182.0   
3565         2022             2503    1998-06-11         180.0   
3566         2023              383    2000-11-03         171.0   
3567         2023              533    1993-01-28         194.0   

      market_value_in_eur  highest_market_value_in_eur  \
0                400000.0                    4000000.0   
1               1000000.0                     750000.0   
2               1000000.0                    1850000.0   
3                400000.0                    4000000.0   
4               1000000.0                    3000000.0   
...                   ...                          ...   
3563             300000.0                     600000.0   
3564            1200000.0                    1200000.0   
3565            4000000.0                    4000000.0   
3566           10000000.0                   30000000.0   
3567            2000000.0                   15000000.0   

     contract_expiration_date  season  home_club_goals  away_club_goals  \
0                  2024-12-31    2012                1                1   
1                  2025-06-30    2012                6                1   
2                  2025-06-30    2012                0                1   
3                  2024-12-31    2012                1                2   
4                  2025-06-30    2012                3                2   
...                       ...     ...              ...              ...   
3563               2026-06-30    2020                3                2   
3564               2024-12-31    2020                2                0   
3565               2025-06-30    2020                0                5   
3566               2024-06-30    2020                2                1   
3567               2024-06-30    2020                1                1   

      home_club_position  away_club_position  attendance  
0                    6.0                 7.0      9728.0  
1                    1.0                12.0      2995.0  
2                    9.0                 5.0     10806.0  
3                    3.0                10.0      9000.0  
4                    3.0                10.0     14485.0  
...                  ...                 ...         ...  
3563                 3.0                18.0      3200.0  
3564                13.0                16.0      5185.0  
3565                14.0                 1.0     42100.0  
3566                 1.0                 6.0     15000.0  
3567                 5.0                13.0      3200.0

In [129]:
Cat = Final_Data.select_dtypes(include = 'object')
Cat

player_name_x competition_id_x                 name  \
0         Aron Johannsson              DK1      Aron Johannsson   
1           Conor O'Brien              DK1        Conor O'Brien   
2        Clarence Goodson              DK1     Clarence Goodson   
3         Aron Johannsson              ELQ      Aron Johannsson   
4          Sacha Kljestan             BESC       Sacha Kljestan   
...                   ...              ...                  ...   
3563        Joseph Efford              BE1        Joseph Efford   
3564         Chris Durkin              BE1         Chris Durkin   
3565        Reggie Cannon              PO1        Reggie Cannon   
3566         Sergino Dest              NL1         Sergino Dest   
3567  John Anthony Brooks               L1  John Anthony Brooks   

              player_code country_of_birth        sub_position position_x  \
0         aron-johannsson    United States      Centre-Forward     Attack   
1            conor-obrien    United States  Defensive Midfield   Midfield   
2        clarence-goodson    United States         Centre-Back   Defender   
3         aron-johannsson    United States      Centre-Forward     Attack   
4          sacha-kljestan    United States  Attacking Midfield   Midfield   
...                   ...              ...                 ...        ...   
3563        joseph-efford    United States      Centre-Forward     Attack   
3564         chris-durkin    United States  Defensive Midfield   Midfield   
3565        reggie-cannon    United States          Right-Back   Defender   
3566         sergino-dest      Netherlands          Right-Back   Defender   
3567  john-anthony-brooks          Germany         Centre-Back   Defender   

       foot           agent_name                 round home_club_manager_name  \
0     right          CAA Stellar           1. Matchday           Peter Rensen   
1     right            Wasserman           1. Matchday       Lars Sondergaard   
2     right            Wasserman           1. Matchday        Auri Skarbalius   
3     right          CAA Stellar  Second Round 1st leg           Peter Rensen   
4      both            Wasserman                 Final      John van den Brom   
...     ...                  ...                   ...                    ...   
3563  right           TOP Agency           7. Matchday          Hernan Losada   
3564  right          Mega Sports           7. Matchday        Wouter Vrancken   
3565  right  YMU Management Ltd.           2. Matchday           Vasco Seabra   
3566  right        Joes Blakborn           3. Matchday           Erik ten Hag   
3567   left                ROGON           2. Matchday      Christian Streich   

     away_club_manager_name                        stadium  \
0              Kent Nielsen                     Ceres Park   
1                Colin Todd                   Sydbank Park   
2               Troels Bech                Brondby Stadium   
3       Temur Shalamberidze                     Ceres Park   
4                Peter Maes  Constant Vanden Stock Stadion   
...                     ...                            ...   
3563            Nicky Hayen              Olympisch Stadion   
3564           Kevin Muscat                   AFAS Stadion   
3565       Sergio Conceicao               Estádio do Bessa   
3566          Thomas Letsch            Johan Cruijff ArenA   
3567         Oliver Glasner                 Dreisamstadion   

                  referee                               home_club_name  \
0        Michael Svendsen                    Aarhus Gymnastik Forening   
1         Claus Bo Larsen                               Sonderjyske Fc   
2     Lars Christoffersen                                   Brondby IF   
3         Boako Jovanetic                    Aarhus Gymnastik Forening   
4       Laurent Colemonts               Royal Sporting Club Anderlecht   
...                   ...                                          ...   
3563      Wesli De Cremer                     

In [130]:
num_list = list(Num.columns)
num_list

['date_x',
 'yellow_cards',
 'red_cards',
 'goals',
 'assists',
 'minutes_played',
 'last_season',
 'current_club_id',
 'date_of_birth',
 'height_in_cm',
 'market_value_in_eur',
 'highest_market_value_in_eur',
 'contract_expiration_date',
 'season',
 'home_club_goals',
 'away_club_goals',
 'home_club_position',
 'away_club_position',
 'attendance']

In [131]:
cat_list = list(Cat.columns)
cat_list

['player_name_x',
 'competition_id_x',
 'name',
 'player_code',
 'country_of_birth',
 'sub_position',
 'position_x',
 'foot',
 'agent_name',
 'round',
 'home_club_manager_name',
 'away_club_manager_name',
 'stadium',
 'referee',
 'home_club_name',
 'away_club_name',
 'aggregate',
 'competition_type']

In [132]:
table = [num_list, cat_list]
print(tabulate({"Categorical Column": cat_list, "Numerical Column": num_list}, 
               headers = ["Categorical Column","Numerical Column"]))

Categorical Column      Numerical Column
----------------------  ---------------------------
player_name_x           date_x
competition_id_x        yellow_cards
name                    red_cards
player_code             goals
country_of_birth        assists
sub_position            minutes_played
position_x              last_season
foot                    current_club_id
agent_name              date_of_birth
round                   height_in_cm
home_club_manager_name  market_value_in_eur
away_club_manager_name  highest_market_value_in_eur
stadium                 contract_expiration_date
referee                 season
home_club_name          home_club_goals
away_club_name          away_club_goals
aggregate               home_club_position
competition_type        away_club_position
                        attendance


## CONNECTING TO MYSQL

In [133]:
import mysql.connector as sql

In [134]:
from sqlalchemy import create_engine

In [136]:
db = sql.connect(host = "localhost", user = "root", password = "yashuanbu1602", database = "school")

In [137]:
mycursor=db.cursor()

In [138]:
mycursor.execute("create database Football_Data")

In [139]:
engine = create_engine("mysql+pymysql://{user}:{pw}@{host}/{db}".
                       format(host='localhost', db='Football_Data', user='root', pw='yashuanbu1602'))

In [140]:
Final_Data.to_sql("Merged_Cleaned_Data", engine,index = False)

3568

In [141]:
Final_Data.to_excel('C8_Football_Data_Preprocessed_Merged_&Cleaned.xlsx', index=False)